Mathlib/Order/Basic.lean

```diff
@@ -676,6 +675,97 @@ instance Order.Preimage.decidable {α β} (f : α → β) (s : β → β → Pro
+    dite_congr rfl (fun _ => rfl) (fun _ => dif_neg h)
+
+@[simp]
+lemma ltByCases_ne (h : x ≠ y) {h₁ : x < y → P} {h₂ : x = y → P} {h₃ : y < x → P}
+    (p : ¬ x < y → y < x := fun h' => h.lt_or_lt.resolve_left h') :
+    ltByCases x y h₁ h₂ h₃ = if h' : x < y then h₁ h' else h₃ (p h') :=
+    dite_congr rfl (fun _ => rfl) (fun _ => dif_pos _)
+
+lemma ltByCases_comm {h₁ : x < y → P} {h₂ : x = y → P} {h₃ : y < x → P}
+    (p : y = x → x = y := fun h' => h'.symm) :
+    ltByCases x y h₁ h₂ h₃ = ltByCases y x h₃ (h₂ ∘ p) h₁ := by
+    refine ltByCases x y (fun h => ?_) (fun h => ?_) (fun h => ?_)
+    · rw [ltByCases_lt h, ltByCases_gt h]
+    · rw [ltByCases_eq h, ltByCases_eq h.symm, comp_apply]
+    · rw [ltByCases_lt h, ltByCases_gt h]
+
+lemma eq_iff_eq_of_lt_iff_lt_of_gt_iff_gt {x' y' : α}
+    (ltc : (x < y) ↔ (x' < y')) (gtc : (y < x) ↔ (y' < x')) :
+    x' = y' ↔ x = y := by simp_rw [eq_iff_le_not_lt, ← not_lt, ltc, gtc]
+
+lemma ltByCases_congr (x' y' : α) {h₁ : x < y → P} {h₂ : x = y → P} {h₃ : y < x → P}
+    {h₁' : x' < y' → P} {h₂' : x' = y' → P} {h₃' : y' < x' → P} (ltc : (x < y) ↔ (x' < y'))
+    (gtc : (y < x) ↔ (y' < x')) (hh'₁ : ∀ (h : x' < y'), h₁ (ltc.mpr h) = h₁' h)
+    (hh'₂ : ∀ (h : x' = y'), h₂ ((eq_iff_eq_of_lt_iff_lt_of_gt_iff_gt ltc gtc).mp h) = h₂' h)
+    (hh'₃ : ∀ (h : y' < x'), h₃ (gtc.mpr h) = h₃' h) :
+    ltByCases x y h₁ h₂ h₃ = ltByCases x' y' h₁' h₂' h₃' := by
+    refine ltByCases x' y' (fun h => ?_) (fun h => ?_) (fun h => ?_)
+    · rw [ltByCases_lt h, ltByCases_lt (ltc.mpr h), hh'₁]
+    · rw [ltByCases_eq h, ltByCases_eq ((eq_iff_eq_of_lt_iff_lt_of_gt_iff_gt ltc gtc).mp h), hh'₂]
+    · rw [ltByCases_gt h, ltByCases_gt (gtc.mpr h), hh'₃]
+
+/-- Perform a case-split on the ordering of `x` and `y` in a decidable linear order,
+non-dependently. -/
+def switchThree (x y : α) (p q r : P) := ltByCases x y (fun _ => p) (fun _ => q) (fun _ => r)
+
+variable {p q r : P}
+
+@[simp]
+lemma switchThree_lt (h : x < y) : switchThree x y p q r = p := ltByCases_lt h
+
+@[simp]
+lemma switchThree_gt (h : y < x) : switchThree x y p q r = r := ltByCases_gt h
+
+@[simp]
+lemma switchThree_eq (h : x = y) : switchThree x y p q r = q := ltByCases_eq h
+
+@[simp]
+lemma switchThree_not_lt (h : ¬ x < y) :
+    switchThree x y p q r = if y < x then r else q := ltByCases_not_lt h
+
+@[simp]
+lemma switchThree_not_gt (h : ¬ y < x) :
+    switchThree x y p q r = if x < y then p else q := ltByCases_not_gt h
+
+@[simp]
+lemma switchThree_ne (h : x ≠ y) : switchThree x y p q r = if x < y then p else r := ltByCases_ne h
+
+lemma switchThree_comm : switchThree x y p q r = switchThree y x r q p := ltByCases_comm
+
+lemma switchThree_congr (x' y' : α) {p' q' r' : P} (ltc : (x < y) ↔ (x' < y'))
+    (gtc : (y < x) ↔ (y' < x')) (hh'₁ : x' < y' → p = p')
+    (hh'₂ : x' = y' → q = q') (hh'₃ : y' < x' → r = r') :
+    switchThree x y p q r = switchThree x' y' p' q' r' := ltByCases_congr _ _ ltc gtc hh'₁ hh'₂ hh'₃
```

Comment:
What's the point in having a separate definition for the non-dependent case? Did you consider making it an `abbrev` so that at least the lemmas don't have to be repeated?

Reply:
Also, I would call it `ltTrichotomy` to mirror `lt_trichotomy`.

Reply:
Well, we do it for ite (this is meant to be the analogy as ite is to dite). Similarly it's sometimes useful to have ergonomic versions of lemmas for that kind of thing? Again, there is precedent for this.

Reply:
Agreed that ltTrichotomy would be a good name

Reply:
I have changed it to `ltTrichotomy`,and made it an `abbrev`, but I still think there is a use for the lemmas (notably `simp` can't solve them - you'd have to unfold the definition). And the abbreviation is useful!